Spork: unfork Blockly

cwillisf opened the pull request:
### Proposed Changes

Use the new version of `scratch-blocks`, which now depends on Blockly instead of being a fork of Blockly.

### Reason for Changes

As stated in scratchfoundation/scratch-blocks#3338:

> We can use new Blockly features!
> We can more easily maintain scratch-blocks!
> We can keep up to date with new Blockly releases!
> We can hang with the cool kids at Blockly-related parties!

### Test Coverage

Many existing tests at least partially cover these changes. We'll also be testing these changes through several manual testing sessions.

adroitwhiz committed:
Replace text-encoding polyfill

gonfunko committed:
fix: temporarily disable functionality in scratch-gui for compatibility with patched scratch-blocks (#1)

gonfunko committed:
fix: run stacks in response to click events (#1)

* fix: run stacks in response to stack click events

* refactor: listen for regular click events to run block stacks

gonfunko committed:
fix: show the correct toolbox based on sprites or the stage being selected (#4)

gonfunko committed:
fix: modify inject options to reflect Scratch behaviors (#5)

gonfunko committed:
fix: add support for Scratch-style procedures (#6)

* fix: add support for Scratch-style procedures

* refactor: remove underscore procedure creation callback

gonfunko committed:
fix: reenable the Scratch colour eyedropper

gonfunko committed:
fix: patch the getCheckboxState method in the new flyout (#7)

gonfunko committed:
fix: select extension categories when added (#8)

gonfunko committed:
fix: Use toolboxitemid instead of id as the identifier attribute for toolbox categories (#9)

gonfunko committed:
fix: use toolboxitemid instead of id as the identifier attribute for extensions toolbox categories (#2)

gonfunko committed:
fix: preserve toolbox scroll position when switching between sprites/the stage (#10)

gonfunko committed:
fix: call reportValue on the ScratchBlocks module instead of the workspace (#11)

gonfunko committed:
fix: call the new glow methods on ScratchBlocks (#12)

* fix: call the new glow methods on ScratchBlocks

* refactor: remove the glowBlock calls

gonfunko committed:
fix: adjust key event filtering to fix the when key pressed block (#13)

gonfunko committed:
refactor: simplify toolbox refreshing behavior (#14)

gonfunko committed:
fix: handle modern workspace comment events (#3)

* fix: handle modern workspace comment events

* fix: correctly access coordinates on events

gonfunko committed:
fix: allow typing into comments (#15)

gonfunko committed:
fix: handle new custom block comment events (#4)

gonfunko committed:
feat: plumb Scratch variable support into the UI (#16)

* chore: format blocks.jsx

* feat: plumb Scratch variable support into the UI

gonfunko committed:
fix: specify the function to be used for prompting about variables (#17)

gonfunko committed:
fix: update the toolbox in response to procedure deletion/creation (#18)

gonfunko committed:
chore: don't specify the renderer/theme (#20)

gonfunko committed:
fix: only refresh the toolbox when procedures are created via undo (#19)

* fix: only refresh the toolbox when procedures are created via undo

* chore: add comment clarifying procedure block creation handling

BeksOmega committed:
fix: add pinch to zoom support (#21)

gonfunko committed:
fix: make dropdown menu shadow block colors consistent (#22)

* chore: format blocks.js

* fix: define shadow block colors consistently with scratch-blocks

* refactor: clean up sound menu recording handler

gonfunko committed:
fix: add the hat extension to extension hat blocks (#5)

* chore: format runtime.js

* fix: add the hat extension to extension hat blocks

gonfunko committed:
fix: add the monitor block extension to extension monitor blocks (#6)

gonfunko committed:
refactor: use block styles instead of colors (#7)

gonfunko committed:
refactor: use block styles instead of directly specifying block colors (#23)

* chore: format monitor.jsx

* chore: format define-dynamic-block.js

* chore: format make-toolbox-xml.js

* chore: format blockHelpers.js

* chore: format theme definitions

* chore: format custom-procedures.jsx

* fix: use styles instead of colors for coloring dynamic blocks

* refactor: define themes in the format expected by Blockly for BlockStyles

* refactor: convert and inject Scratch themes into Blockly

* chore: add comments clarifying the color/colour dichotomy

gonfunko committed:
refactor: improve efficiency of toolbox updates (#24)

gonfunko committed:
fix: partially roll back flyout optimization (#25)

gonfunko committed:
fix: update VM state when blocks are removed from an input previously occupied by a shadow block (#8)

* chore: format engine/blocks.js

* fix: restore shadow blocks when removing a block from an input

gonfunko committed:
fix: prevent exception when switching languages (#27)

gonfunko committed:
fix: fix bug that prevented displaying the procedure editor modal on mobile (#26)

gonfunko committed:
fix: avoid clearing the state of the sensing_of block by refreshing the toolbox (#28)

gonfunko committed:
fix: fix bug that could result in the VM's representation of shadow blocks getting into a bad state (#9)

cwillisf committed:
style: run "eslint --fix ."

gonfunko committed:
refactor: fix compatibility with checkboxes and category status indicators (#29)

* refactor: fix compatibility with checkboxes and category status indicators

* chore: remove errant logging

cwillisf committed:
ci: enable alpha and beta releases

cwillisf committed:
fix: use scratch-blocks@^2.0.0-beta

cwillisf committed:
chore(deps): update scratch-blocks for unforking test

cwillisf committed:
fix(release): release beta branch under beta label(?)

cwillisf committed:
ci: make temporary "spork" release channel

cwillisf committed:
chore(deps): update deps for spork test

cwillisf committed:
chore(deps): update deps for spork test

gonfunko committed:
fix: Fix test failures. (#10)

gonfunko committed:
fix: Fix test failures. (#10)

cwillisf committed:
chore(deps): update deps for spork test

cwillisf committed:
chore(deps): update deps for spork test

cwillisf committed:
style: re-apply existing lint rules

cwillisf committed:
fix: update "colour" property names for dark theme and mocks

cwillisf committed:
test: update dynamic block tests for colour -> style

cwillisf committed:
test: fix theme injection tests

cwillisf committed:
build: fix webpack-dev-server configuration

The playgrounds themselves are still broken, but it's a step...

gonfunko committed:
fix: fix preserving toolbox scroll position for new continuous toolbox plugin (#30)

cwillisf committed:
chore(deps): update deps for spork test

cwillisf committed:
chore(deps): update deps for spork test

cwillisf committed:
style: lint fixes

cwillisf committed:
Revert "fix: adjust key event filtering to fix the when key pressed block (#13)"

This reverts commit 87a9787cf5e977dd36bf77354ec4ca397bcc2969.

cwillisf committed:
test: update XPath expressions for Blockly v12 beta

cwillisf committed:
test: move Selenium debugging info to files in test-results/

Port 554c82efdaf0bb93fd82730ba0260c863c9f1d70 from the monorepo and
apply a minor supporting `package.json` change

cwillisf committed:
style: copy style changes from gonfunko/modern-blockly

cwillisf committed:
Merge branch 'develop' into spork

cwillisf committed:
Merge branch 'develop' into spork

gonfunko committed:
fix: use derived IDs for block comments (#11)

cwillisf committed:
test: update more XPath expressions for Blockly v12

cwillisf committed:
fix: variables got shy when flipping between code and project view

Reverts one line of 9703bc6531e101d7684ebb394f5afacb287f36fd

cwillisf committed:
Merge branch 'develop' into spork

cwillisf committed:
Merge branch 'build-script-improvements' into spork

cwillisf committed:
test: avoid false positives in xpath @class tests

cwillisf committed:
Merge branch 'develop' into spork

cwillisf committed:
Merge branch 'develop' into spork

cwillisf committed:
Merge branch 'develop' into spork

cwillisf committed:
ci: fix(?) pushing from the publish workflow

cwillisf committed:
ci: move commitlint into its own separate workflow

This way it doesn't prevent other tests from running

cwillisf committed:
ci: try again to get the push to work

cwillisf committed:
ci: temporarily disable publishing to NPM

cwillisf committed:
ci: determine tag for "npm publish" from branch name

cwillisf committed:
ci: temporarily disable pre-build/test in publish workflow

cwillisf committed:
ci: make release commit message pass commitlint

cwillisf committed:
ci: re-enable publish steps that were disabled for testing purposes

cwillisf committed:
ci: publish to NPM under a tag determined by branch

cwillisf committed:
Merge branch 'develop' into spork

cwillisf commented:
Status update for anyone who might be watching this PR:

I know it looks like not much is happening here, but gears are turning behind the scenes. We're working on testing, and I'm sure there will be things to fix before we merge this, but I believe we're getting close. If anyone in the community wants to try this branch, please comment here if you find problems! Other than the zoom icons -- we know about that one :laughing:

Thank you!

If all goes as planned, I'd like to merge this sometime this summer, specifically before North American schools get going again for fall terms. If we get lucky and it's even earlier, that's cool too!

Some of our friends at Google haven't signed the CLA, but that's OK for now. We're working through some paperwork, but we're all friends, and I'm not worried about it. It's just slow; I apologize for that.

Samq64 commented:
On `scratch-gui`'s `spork-test` branch custom block inputs get cleared when editing the definition (https://github.com/gonfunko/scratch-blocks/issues/228).

mxmou commented:
> If anyone in the community wants to try this branch, please comment here if you find problems!

If you make the following script and then click the green flag, nothing happens:
![message1](https://github.com/user-attachments/assets/cd09b111-6330-4058-a920-bb2106c0829c)

Switching to another sprite and then back to the first fixes it. This seems to happen because the VM receives a `var_delete` event for the message1 broadcast when a broadcast block is dragged onto the workspace. I wasn't able to find the exact cause of this, but it's related to potential variables.

Another issue is that the `isLocal` property of variables gets reset to false when switching sprites, which causes Scratch to think variable names conflict when they don't. To reproduce:
1. Make a project with two sprites.
2. In the first sprite, make a variable for this sprite only named "a".
3. In the second sprite, make a variable for this sprite only named "b".
4. Switch to the first sprite and back to the second.
5. Rename "b" to "a".

An error message will appear saying that "a" already exists, even though it's in another sprite. This doesn't happen if step 4 is omitted.

There are also some UI bugs:
* Up & down arrow keys don't work in context menus and dropdowns.
* Scrolling in long dropdowns (for example the "switch costume" dropdown if the sprite has many costumes) doesn't work very well. I can use the scrollbar and mouse wheel, but when I move the mouse after scrolling, it scrolls to the bottom instead of staying where it was.
* The lines on the direction picker are white instead of black in high contrast mode, which makes them hard to see:
![high_contrast_direction_picker](https://github.com/user-attachments/assets/7c1b229a-1fca-4cbc-8ce3-ab055d86ae34)

cwillisf committed:
ci: publish all builds to GH Pages

cwillisf committed:
ci: clean up stale GH Pages subdirectories

mxmou commented:
Another issue: block comments are attached to the wrong side of the block in RTL.

![image](https://github.com/user-attachments/assets/67a1e2a4-390e-41a4-a88c-59c58390c1bb)

Samq64 commented:
When dragging a custom block definition onto a stack the shadow is not resized to the correct width:
<img width="165" height="143" alt="small-block-shadow" src="https://github.com/user-attachments/assets/bedcba58-e2dc-4f42-8d6c-5436ff757405" />

gonfunko committed:
fix: use serialization wrapper (#32)

gonfunko committed:
fix: plumb `FieldNote` through to the VM (#33)

gonfunko commented:
Thanks for all the comments and bug reports! At this time, I believe everything reported here has been resolved in my fork; feel free to report any more issues you encounter in [my repo](https://github.com/gonfunko/scratch-blocks/issues), I very much appreciate it.

cwillisf committed:
fix(deps): update to scratch-blocks@2.0.0-spork.5

cwillisf committed:
Merge branch 'develop' into spork

cwillisf committed:
style: use single quotes to fix lint

cwillisf committed:
test: update comment ID in engine adapter test

cwillisf committed:
ci: de-matrix GH Pages deploy

cwillisf committed:
ci: skip GH Pages publish from fork PRs

cwillisf committed:
ci: output more info for debugging GHA

zfangqijun commented:
@cwillisf hello，is there a preview page for spork?
I am developing using the spork branch and encountering some bugs, I want to compare and see if it's a problem with my own code.

Samq64 commented:
Will Google [transferring Blocky to the Raspberry Pi Foundation](https://opensource.googleblog.com/2025/10/blockly-graduates-from-google.html) impact this?

> is there a preview page for spork?

Yes, https://scratchfoundation.github.io/scratch-editor/spork/scratch-gui/ although I am not sure if it's fully up-to-date.

zfangqijun commented:
> Will Google [transferring Blocky to the Raspberry Pi Foundation](https://opensource.googleblog.com/2025/10/blockly-graduates-from-google.html) impact this?
> 
> > is there a preview page for spork?
> 
> Yes, https://scratchfoundation.github.io/scratch-editor/spork/scratch-gui/ although I am not sure if it's fully up-to-date.

thanks！🌹

cwillisf commented:
Hey all, thanks for checking in!

> Yes, https://scratchfoundation.github.io/scratch-editor/spork/scratch-gui/ although I am not sure if it's fully up-to-date.

That should always be up to date with what's in this branch, as it's automatically deployed as part of the CI/CD scripts in this repository. However, this branch is somewhat out of date right now: I need to merge the React 18 and Face Sensing changes from `develop` into this branch.

cwillisf commented:
> Will Google [transferring Blocky to the Raspberry Pi Foundation](https://opensource.googleblog.com/2025/10/blockly-graduates-from-google.html) impact this?

I'm excited for the Blockly team! The Scratch team is friends with the folks at the Raspberry Pi Foundation, and we've worked together in the past. We're all committed to seeing this through and continuing to work together in the future.

TL;DR: all good! :tada:

cwillisf committed:
Merge remote-tracking branch 'origin/develop' into spork

cwillisf committed:
chore(deps): update deps for spork test

cwillisf committed:
Merge remote-tracking branch 'origin/develop' into spork

cwillisf committed:
style: fix scratch-vm lint

cwillisf committed:
style: match formatting in develop to reduce diff noise

cwillisf committed:
fix: resolve oops on dragging code to backpack

cwillisf committed:
fix(deps): update dependency scratch-blocks to v2.0.0-spork.8

Samq64 commented:
I have noticed the reset position button also appears in the top-left corner of the make a block modal and sometimes all 3 buttons are in the bottom right of the modal instead.
<img width="425" height="197" alt="image" src="https://github.com/user-attachments/assets/9f3a5d29-4d64-41a4-9091-20359e969cf6" />

cwillisf commented:
@Samq64, thanks for reporting that! I'm having trouble getting any of the buttons to appear in the bottom-right corner of the "Make a Block" dialog. Do you know of a set of steps that's more likely to cause that behavior?

On the other hand, I can easily repro the issue with the buttons stacked on top of each other in the upper left corner, so I'll start working on that and might fix the other along the way.

cwillisf committed:
fix(deps): update dependency scratch-blocks to v2.0.0-spork.9

cwillisf committed:
fix(deps): adapt to modernized scratch-blocks spork build output

Specifics:

- update linting to handle indirect TS imports (`export * from ...`)
  - update to `eslint-config-scratch@^13`
  - use `eslint-plugin-import-x` instead of `eslint-plugin-import`
  - remove workarounds for issues resolved by the above
  - add a rule to ignore webpack inline loader syntax
- update to `scratch-blocks@2.0.0-spork.10`
  - import `* as ScratchBlocks` instead of `{ScratchBlocks}`
    - improves tree-shaking support
  - don't hard-code `scratch-blocks` output path in `scratch-vm`'s
    wepack configuration

cwillisf committed:
fix(deps): fix build with scratch-blocks@2.0.0-spork.11

Give up (for now) on externalized Blockly

cwillisf committed:
fix: don't try to load sounds in 'Make a Block' dialog

cwillisf committed:
Merge branch 'develop' into spork

cwillisf committed:
Merge branch 'develop' into spork

cwillisf committed:
test: fix flyout block scopes

cwillisf committed:
Merge branch 'develop' into spork

cwillisf committed:
fix(deps)!: update scratch-blocks to v2.0.0

BREAKING CHANGE: Updating to `scratch-blocks@^2` and Blockly v12 brings
significant changes to the document structure and build system. Most
applications depending on `scratch-gui` will likely still work, but it's
possible that some of these changes will be disruptive to more sensitive
applications. It's also possible, maybe likely, to affect any tests that
interact with the blocks workspace or wrappers around it.

cwillisf commented:
Contributions here from the Blockly team are covered by a separate agreement, so the Signature Assistant failure is OK in this specific case.

cwillisf reviewed (commented):
We're aware of a few remaining issues, but it's time to merge this into our primary development line. 🥳
